Scripts/DowloadPortalItemData.swift

```diff
@@ -242,45 +228,55 @@ let previousDownloadedItems: DownloadedItems = {
-        continue
-    }
-    
-    let destinationURL = downloadDirectoryURL.appendingPathComponent(portalItem.identifier, isDirectory: true)
-    
-    // Deletes the directory when the item is not in the plist.
-    try? FileManager.default.removeItem(at: destinationURL)
-    
-    do {
-        // Creates an enclosing directory with portal item ID as its name.
-        try FileManager.default.createDirectory(at: destinationURL, withIntermediateDirectories: false)
-    } catch {
-        print("error: Error creating download directory: \(error.localizedDescription)")
-        exit(1)
-    }
-    
-    print("note: Downloading item \(portalItem.identifier)")
-    fflush(stdout)
-    dispatchGroup.enter()
-    downloadFile(at: portalItem.dataURL, to: destinationURL) { result in
-        switch result {
-        case .success(let url):
-            downloadedItems[portalItem.identifier] = url.lastPathComponent
-            dispatchGroup.leave()
-        case .failure(let error):
-            print("error: Error downloading item \(portalItem.identifier): \(error.localizedDescription)")
-            URLSession.shared.invalidateAndCancel()
+await withTaskGroup(of: Void.self) { group in
+    for portalItem in portalItems {
+        // Checks to see if an item is already downloaded.
+        guard downloadedItems[portalItem.identifier] == nil else {
+            print("note: Item already downloaded: \(portalItem.identifier)")
+            continue
+        }
+        
+        let destinationURL = downloadDirectoryURL.appendingPathComponent(
+            portalItem.identifier,
+            isDirectory: true
+        )
+        
+        // Deletes the directory when the item is not in the plist.
+        try? FileManager.default.removeItem(at: destinationURL)
+        
+        do {
+            // Creates an enclosing directory with the portal item ID as its name.
+            try FileManager.default.createDirectory(
+                at: destinationURL,
+                withIntermediateDirectories: false
+            )
+        } catch {
+            print("error: Error creating download directory: \(error.localizedDescription)")
             exit(1)
         }
+        
+        print("note: Downloading item \(portalItem.identifier)")
+        fflush(stdout)
+        group.addTask {
+            do {
+                let downloadName = try await downloadFile(
+                    from: portalItem.dataURL,
+                    to: destinationURL
+                )
+                print("note: Downloaded item: \(portalItem.identifier)")
+                fflush(stdout)
+                
+                await MainActor.run {
+                    downloadedItems[portalItem.identifier] = downloadName
+                }
+            } catch {
+                print("error: Error downloading item \(portalItem.identifier): \(error.localizedDescription)")
+                URLSession.shared.invalidateAndCancel()
```

Comment:
The `invalidateAndCancel` cancels all outstanding tasks for this session. Even if one download fails, do you want to cancel all outstanding tasks? Is that the intention?

Reply:
I believe that was the intention in the original script yes. The script exits right after this line to prevent building with a failed download. It is good to invalidate before then.